src/contracts/middleware/BLSPubkeyRegistry.sol

```diff
@@ -0,0 +1,211 @@
+        return quorumToApk[quorumNumber];
+    }
+
+    function globalApk() external view returns (BN254.G1Point memory){
+        return _globalApk;
+    }
+
+
+    /**
+     * @notice get hash of the apk of `quorumNumber` at `blockNumber` using the provided `index`;
+     * called by checkSignatures in BLSSignatureChecker.sol.
+     */
+    function getApkHashForQuorumAtBlockNumberFromIndex(uint8 quorumNumber, uint32 blockNumber, uint256 index) external view returns (bytes32){
+        ApkUpdate memory quorumApkUpdate = quorumToApkUpdates[quorumNumber][index];
+        require(blockNumber >= quorumApkUpdate.updateBlockNumber, "BLSPubkeyRegistry.getApkHashForQuorumAtBlockNumberFromIndex: index too recent");
+
+        if (index != quorumToApkUpdates[quorumNumber].length - 1){
+            require(blockNumber < quorumApkUpdate.nextUpdateBlockNumber, "BLSPubkeyRegistry.getApkHashForQuorumAtBlockNumberFromIndex: not latest apk update");
+        }
+
+        return quorumApkUpdate.apkHash;
+    }
+
+	/**
+     * @notice get hash of the apk among all quourums at `blockNumber` using the provided `index`;
+     * called by checkSignatures in BLSSignatureChecker.sol.
+     */
+    function getGlobalApkHashAtBlockNumberFromIndex(uint32 blockNumber, uint256 index) external view returns (bytes32){
+        ApkUpdate memory globalApkUpdate = globalApkUpdateList[index];
+        require(blockNumber >= globalApkUpdate.updateBlockNumber, "BLSPubkeyRegistry.getGlobalApkHashAtBlockNumberFromIndex: blockNumber too recent");
+
+        if (index != globalApkUpdateList.length - 1){
+            require(blockNumber < globalApkUpdate.nextUpdateBlockNumber, "getGlobalApkHashAtBlockNumberFromIndex.getGlobalApkHashAtBlockNumberFromIndex: not latest apk update");
+        }
+
+        return globalApkUpdate.apkHash;
+    }
+
+    function getQuorumApkHistoryLength(uint8 quorumNumber) external view returns(uint32){
+        return uint32(quorumToApkUpdates[quorumNumber].length);
+    }
+
+    function getGlobalApkHistoryLength() external view returns(uint32){
+        return uint32(globalApkUpdateList.length);
+    }
+
+    function _processGlobalApkUpdate(BN254.G1Point memory newGlobalApk) internal returns(bytes32){
+        bytes32 globalApkHash = BN254.hashG1Point(newGlobalApk);
+        if(globalApkUpdateList.length > 0){
+            globalApkUpdateList[globalApkUpdateList.length - 1].nextUpdateBlockNumber = uint32(block.number);
+        }
+        _globalApk = newGlobalApk;
+
+        ApkUpdate memory latestGlobalApkUpdate;
+        latestGlobalApkUpdate.apkHash = globalApkHash;
+        latestGlobalApkUpdate.updateBlockNumber = uint32(block.number);
+        globalApkUpdateList.push(latestGlobalApkUpdate);
+
+        return globalApkHash;
+    }
+
+    function _processQuorumApkUpdate(uint8[] memory quorumNumbers, BN254.G1Point memory pubkey, bool isRegistration) internal {
```

Comment:
i feel like a cleaner method of doing this is renaming `pubkey` to `point` and just negating the pubkey and inputting it during deregistration and keeping it as the pubkey during registration, and removing the `isRegistration` param. This abstracts the most logic and makes this unit simpler.